torch/optim/asgd.py

```diff
@@ -428,16 +428,7 @@ def _multi_tensor_asgd(
             torch._foreach_copy_(grouped_etas, new_etas)
         else:
             step = grouped_state_steps[0].item()
```

Comment:
Thanks for the PR. You're right that this optimization would be better than the current code, but you've just brought to my attention that the current code is wrong because it makes the ill assumption that every param will have the same step. This is not true! If certain params are frozen in an iteration, the step counts will vary among different params, which is why we store a step per param in the first place.

So we should actually fix this code to not index on step being the same value and use the forloop properly to access step. We should also add a test case in test/test_optim.py under TestOptimRenewed where the steps vary across params. Would you be interested in doing this instead?

Reply:
Sure, I would like to fix it.

Reply:
Feel free to update this PR and ping me when it’s ready!